### PR TITLE
Add os_disk_encryption_set to azure_rm_virtualmachine

### PR DESCRIPTION
**Pull request by ephracis**
##### SUMMARY
Add the parameter `os_disk_encryption_set` to the
`azure_rm_virtualmachine` module, making it possible to specify which DES to use when encrypting the OS disk for a newly created VM.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
azure_rm_virtualmachine

##### ADDITIONAL INFORMATION

This is required when deploying a VM into one landing zone, using an encrypted Compute Gallery Image from another landing zone. If this is not set it will try to use the DES of the image which will fail since it belongs to another LZ.

### CODE REVIEW

File: plugins/modules/azure_rm_virtualmachine.py [hunk 1]
```diff
@@ -206,6 +206,9 @@
     os_disk_size_gb:
         description:
             - Size of OS disk in GB.
+    os_disk_encryption_set:
+        description:
+            - ID of disk encryption set for OS disk.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py:234]: Missing type
```suggestion
            - ID of disk encryption set for OS disk.
        type: str
```

**Comment by ephracis** [comment]
I added some integration tests to use the new disk encryption which works fine.

File: tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
+
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml:4]: delete blank lines
```suggestion
```

File: tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    key_url: "{{ results['keys'][0]['kid'] }}"
+
+- name: Create disk encryption set
+  azure_rm_diskencryptionset:
+    resource_group: "{{ resource_group }}"
+    name: "{{ des_name }}"
+    source_vault: "myvault{{ des_name }}"
+    key_url: "{{ key_url }}"
+    state: present
+  register: des_results
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml]: No new line character at the end of file
```suggestion
  register: des_results
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 4]
```diff
@@ -1559,6 +1565,11 @@ def exec_module(self, **kwargs):
                             vm_dict['os_profile']['linux_configuration']['disable_password_authentication']:
                         self.fail("(PropertyChangeNotAllowed) Changing property 'linuxConfiguration.disablePasswordAuthentication' is not allowed.")
 
+                current_os_des_id = vm_dict['storage_profile'].get('os_disk', {}).get('managed_disk', {}).get('disk_encryption_set', {}).get('id', {})
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                current_os_des_id = vm_dict['storage_profile'].get('os_disk', {}).get('managed_disk', {}).get('disk_encryption_set', {}).get('id', None)
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 5]
```diff
@@ -1559,6 +1565,11 @@ def exec_module(self, **kwargs):
...
+                if self.os_disk_encryption_set is not None and current_os_des_id is not None:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py:1571]:
```suggestion
                if self.os_disk_encryption_set is not None and current_os_des_id is not None and self.os_disk_encryption_set != current_os_des_id:
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 6]
```diff
@@ -1172,6 +1177,7 @@ def __init__(self):
         self.os_type = None
         self.os_disk_caching = None
         self.os_disk_size_gb = None
+        self.os_disk_encryption_set = None
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py:1180]: Please add required_if = [('os_disk_encryption_set', '*', ['managed_disk_type'])] and add "required_if=required_if" to line 1219

**Comment by Fred-sun** [comment]
> I added some integration tests to use the new disk encryption which works fine.

@ephracis  Yes, I see. But os_disk only works if managed_disk_type is not None（Please check lines 1704 through 1711）, and managed_disk_type is not specified in your use case, so it works fine. Thank you!

File: plugins/modules/azure_rm_virtualmachine.py (hunk 1 above)
Comment (ephracis) [reply plugins/modules/azure_rm_virtualmachine.py:234]: Done

File: tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml (hunk 2 above)
Comment (ephracis) [reply tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml:4]: Done

File: tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml (hunk 3 above)
Comment (ephracis) [reply tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml]: Done

File: plugins/modules/azure_rm_virtualmachine.py (hunk 4 above)
Comment (ephracis) [reply plugins/modules/azure_rm_virtualmachine.py]: Done

File: plugins/modules/azure_rm_virtualmachine.py (hunk 6 above)
Comment (ephracis) [reply plugins/modules/azure_rm_virtualmachine.py:1180]: Done

**Comment by ephracis** [comment]
> > I added some integration tests to use the new disk encryption which works fine.
> 
> @ephracis Yes, I see. But os_disk only works if managed_disk_type is not None（Please check lines 1704 through 1711）, and managed_disk_type is not specified in your use case, so it works fine. Thank you!

Good catch! I've added `managed_disk_type: Standard_LRS` to that specific test case but then I get this error:

> KeyVaultAccessForbidden. Please grant get, wrap and unwrap key permissions to disk encryption set.

Unfortunately I don't see any Ansible module for doing this. Any suggestions?

**Comment by ephracis** [comment]
Ok, I think I've managed to get the integration tests running. The problem was that I had to add the DES to the access list.

Then I ran into a second issue where I had to enable purge protection in order to use the KV for disk encryption. This means that all KVs will reside deleted but unpurged for 90 days. So in order to not get naming conflicts with these keys, they are given a date-based suffix.

Let me know if there's anything else that needs fixing. 😅

**Review by xuzhang3 — APPROVED** [review]

**Comment by xuzhang3** [comment]
@ephracis  LGTM

**Pull request merged**